src/osp/Resource/Resource.h

```diff
@@ -24,29 +24,37 @@
  */
 #pragma once
 
+#include <assert.h>
 #include <map>
+#include <optional>
 #include <string>
 
 namespace osp
 {
 
+struct construct_tag {};
+struct reserve_tag {};
+
 template <class TYPE_T>
 struct Resource
 {
-    Resource(bool loaded)
+    Resource(reserve_tag)
      : m_data()
-     , m_loaded(loaded)
     { }
-    Resource(bool loaded, TYPE_T&& data)
+    Resource(construct_tag)
+     : m_data(TYPE_T())
+    { }
+    Resource(construct_tag, TYPE_T&& data)
```

Comment:
This is my first time using a tag dispatch; I kind of don't know what I'm doing. `Resource()` became ambiguous, which can either refer to default-constructing TYPE_T, or initializing std::optional as null.

Reply:
I recommend collapsing the two construct_tag constructors into one, with TYPE_T&& data = {}.

Reply:
I'll combine the constructors, but I'll keep the tags as-is. I'm not sure if I fully understand what you mean, but I feel that parameter pack would end up moving construct-related compiler error messages deep into Resource, which I found caused headaches. Isn't nullopt_t already a tag? The terms reserve and construct are also more explicit to this application, and makes sure the first argument has a consistent meaning.

Reply:
std::nullopt_t is natively understood by std::optional to not construct anything.

You're basically just reinventing functionality that std::optional already has built in.

Reply:
This is unlikely, but if one of TYPE_T's constructors accepts a std::nullopt_t, then it would become ambiguous whether Resource(std::nullopt_t()) should construct or not. When there's two separate tags for construct and reserve, the first argument will never be ambiguous.

Reply:
You would out the std::nullopt inside of an initializer list, so that the std::optional didn't pick it out.

Reply:
In this rare case, Package::add causes unintended behavior of constructing the optional as empty.
```cpp
package.add<AcceptsNullopt>("name", std::nullopt);
// calls Resource<AcceptsNullopt>(std::nullopt_t)
```

Needing an initializer list specifically for this case is inconsistent.
```cpp
package.add<AcceptsNullopt>("name", {std::nullopt});
```

Anyways, it turns out stl already dealt with this issue with two separate tags. Is this good? 
```cpp
template <class TYPE_T>
struct Resource
{
    Resource(std::nullopt_t)
     : m_data(std::nullopt)
    { }
    template<typename ... ARGS_T>
    Resource(std::in_place_t, ARGS_T&& ... args)
     : m_data(std::in_place, std::forward<ARGS_T>(args)...)
    { }
    Resource(Resource&& move) = default;
    Resource(const Resource& copy) = delete;
    ~Resource()
    {
        assert(m_refCount == 0);
    }

    std::optional<TYPE_T> m_data;

    int m_refCount{0};
};
```